File: leetcode/golang/LCR_170_shu_zu_zhong_de_ni_xu_dui_lcof.go
func mergeSort(record []int, tmp []int, left, right int) int {
    if left >= right {
        return 0
    }
    mid := left + (right - left) / 2
    res := mergeSort(record, tmp, left, mid) + mergeSort(record, tmp, mid+1, right)
    for k := left; k <= right; k++ {
        tmp[k] = record[k]
    }
    leftStart, rightStart := left, mid+1
    for k := left; k <= right; k++ {
        if leftStart == mid+1 {
            record[k] = tmp[rightStart]
            rightStart++
        } else if rightStart == right+1 || tmp[leftStart] <= tmp[rightStart] {
            record[k] = tmp[leftStart]
            leftStart++
        } else {
            record[k] = tmp[rightStart]
            rightStart++
            res += mid - leftStart + 1
        }
    }
    return res;
}

func reversePairs(record []int) int {
    tmp := make([]int, len(record))
    return mergeSort(record, tmp, 0, len(record)-1)
}

File: leetcode/c++/LCR_170_shu_zu_zhong_de_ni_xu_dui_lcof.cpp
// https://leetcode.cn/problems/shu-zu-zhong-de-ni-xu-dui-lcof

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int mergeSort(vector<int>& record, vector<int>& tmp, int l, int r) {
        if (l >= r)
            return 0;
        int m = l + (r-l) / 2;
        int res = mergeSort(record, tmp, l, m) + mergeSort(record, tmp, m+1, r);
        for (int k = l; k <= r; k++) 
            tmp[k] = record[k];
        int left_start = l, right_start = m+1;
        for (int k = l; k <= r; k++) {
            if (left_start == m+1)
                record[k] = tmp[right_start++];
            else if (right_start == r+1 || tmp[left_start] <= tmp[right_start])
                record[k] = tmp[left_start++]; // 稳定排序
            else {
                record[k] = tmp[right_start++];
                res += m-left_start+1;
            }
        } 
        return res;
    }
    int reversePairs(vector<int>& record) {
       vector<int> tmp(record.size(), 0);
       return mergeSort(record, tmp, 0, record.size() - 1);
    }
};

// 用归并排序来避免超时，在合并的时候统计逆序对，res += m-left_start+1这个很关键，因为左边都是排序好了的
// 只要比左边起始位置小，那么说明比左边所有剩余数都小
// 时间复杂度为O(nlogn)，空间复杂度为O(n)